Galaxus.Functional/Linq/OptionLinqExtensions.cs

```diff
@@ -56,8 +52,15 @@ public static Option<TSource> FirstOrNone<TSource>(this IEnumerable<TSource> sou
         /// <exception cref="ArgumentNullException"><c>source</c> or <c>predicate</c> is <c>null</c>.</exception>
         public static Option<TSource> FirstOrNone<TSource>(this IEnumerable<TSource> source, Func<TSource, bool> predicate)
         {
-            var value = source.Where(predicate).ToList();
-            return value.Count == 0 ? Option<TSource>.None : Option<TSource>.Some(value[0]);
+            foreach (var element in source)
+            {
+                if (predicate(element))
+                {
+                    return Option<TSource>.Some(element);
+                }
+            }
+
+            return Option<TSource>.None;
```

Comment:
Couldn't we just write `source.Where(predicate).Select(Option<TSource>.Some).FirstOrDefault()` instead?

Reply:
You can, and it is actually the current implementation in Funcky [https://github.com/polyadic/funcky/blob/master/Funcky/Extensions/EnumerableExtensions/FirstOrNone.cs]. We tested the two versions against each other, and it looks like the optimization is quite a bit faster.

Reply:
I was able to reproduce your results with my own benchmark (see https://gist.github.com/SamuelBittmann/e9d2487f494e6f4a6db762fb2781c39f). At least when I was using `Enumerable.Range(...)` as the source. When I converted the range to a list first I got inverted results and the Linq implementation won (N is the number of elements in the enumerable):
![image](https://user-images.githubusercontent.com/9946228/95852074-73223380-0d53-11eb-8784-8945182a5bc6.png)
For consistency reasons, and because we at Digitec Galaxus mostly use these methods together with lists, I would suggest we use the Linq version here. Or tell me if I did something wrong with the benchmark 😊

Reply:
Interesting test. Feel free to chose any implementation you see fit. The net gain is neglectable in most scnerios anyway.